### ISSUE TITLE
Ensure that shaders are generated once per entry point

### ISSUE THREAD

**Issue by grovesNL** [gfx-rs/gfx#3435]
rust-gpu outputs vertex and fragment shaders in a single SPIR-V file, but we haven't been testing this path much yet because we usually receive shaders as separate files. This appears to cause problems for the MSL generation path (possibly HLSL and GLSL too but I haven't checked them yet).

Currently it looks like we convert SPIR-V to MSL for one entry point only and then try to reuse the cache, but this fails with missing entry points when they exist in a single file. We can probably include the entry point as part of the cache key to avoid reusing the cache across different entry points for now (i.e. until this is supported in SPIRV-Cross/naga).

For example, if a shader has `main_vs` and `main_fs`, we might compile only one of these paths. So if we compiled/cached the fragment shader entry point, we might end up with `Device::create_render_pipeline)(no module): Function retrieval error "Function \'main_vs\' does not exist"` later, because the `main_vs` entry doesn't exist.

**Comment by NthTensor** [comment]
I stumbled across this problem trying out rust-gpu for mac. The problem is that SPIRV-Cross can only generate MSL with a single endpoint. The C++ lib lets you choose which entrypoint you want, but wrapper doesn't and always outputs the first entry-point it finds. We need to add a `set_entry_point` wrapper to the spirv_cross crate, and then fix `compile_shader_library` in `metal/devise.rs` to use it.

I actually just got this working, and will be setting up some PRs when I next get the chance, probably on monday.
